0000-settlement-engines/0000-settlement-engines.md

```diff
@@ -0,0 +1,395 @@
+
+#### Request
+
+```http
+POST /accounts/:id/settlements HTTP/1.1
+Accept: application/json
+Content-Type: application/json
+Idempotency-Key: <key>
+```
+
+> **[`Quantity`](#quantity-json-type)** to be credited to the account as an incoming settlement
+
+#### Response
+
+```http
+HTTP/1.1 201 CREATED
+Content-Type: application/json
+```
+
+> **[`Quantity`](#quantity-json-type)** credited to the account as an incoming settlement, which MUST  be less than or equal to the quantity in the original request
+
+### Send outgoing message
+
+Send the message to the given peer's settlement engine and return its response. The connector handles [proxying the message](#exchanging-messages) through the peer's connector.
+
+#### Request
+
+```http
+POST /accounts/:id/messages HTTP/1.1
+Accept: application/octet-stream
+Content-Type: application/octet-stream
+```
+
+> _&lt;raw bytes of message&gt;_
+
+#### Response
+
+```http
+HTTP/1.1 200 OK
+Content-Type: application/octet-stream
+```
+
+> _&lt;raw bytes of response&gt;_
+
+## Idempotence
+
+Idempotent requests ensure each side effect only happens once, even though the same request may be called multiple times. For example, if a request to perform a settlement fails due to a network connection error, [idempotence](https://en.wikipedia.org/wiki/Idempotence) prevents a client from accidentally triggering multiple settlements when it retries the request.
+
+### Performing idempotent requests
+
+Requests to settle or requests to credit incoming settlements MUST include an idempotency key, or globally unique string, within an `Idempotency-Key: <key>` header. To avoid collisions, this key MUST be derived from a cryptographically secure source of randomness.
+
+### Recommended retry behavior
+
+If the client receives no response, it MUST retry the request with the same idempotency key.
+
+To prevent overwhelming the server, the client SHOULD exponentially backoff each retry attempt and add random "jitter" to vary the retry interval. After several failed attempts, the client MUST rollback any side effects it performed, such as [refunding balances](#double-entry-bookkeeping).
+
+### Handling idempotent requests
+
+Endpoints to settle and endpoints to credit incoming settlements MUST support idempotency keys.
+
+Before an endpoint responds to the request with a new idempotency key (one it hasn't seen before), it should persist the idempotency key and the state of its response. If a subsequent request is sent with the same idempotency key, the server should use the state from the initial request to return the same response. Servers MUST persist idempotency keys and response state for at least 24 hours after the initial request was performed.
+
+Idempotency keys may be any length, but at a minimum, servers MUST support idempotency keys up to 32 characters long.
```

Comment:
Is this requirement driven by a particular use-case?

Reply:
Two motivations: (a) implementations shouldn't have to support infinite-length idempotency keys, but have different maximum lengths they support, and (b) should support at least using 32-byte hashes or UUIDs as idempotency keys.

Upon re-examination, I'm not sure it's entirely necessary to spec this... implementations can support up to whatever length they want to, and this seems kinda arbitrary.